### ISSUE TITLE
Way for subscriber callback to call class method

### ISSUE THREAD

**Issue by mikepurvis** [ros-drivers/rosserial#116]
Not sure if this is a good idea, but I have need of it, and so am polling the audience:

It'd be convenient to have an overloaded version of Subscriber which lets you pass a class method and instance pointer. Unfortunately, there are a couple factors which make this trickier than the roscpp version of the same behaviour:
- We don't have a "subscribe" method (because it would require dynamic memory allocation), so any additional template arguments to a Subscriber instance have to be declared upfront (no class constructor type deduction), which makes for an awkward Subscriber instantiation.
- We obviously can't depend on boost::bind/function.

One possibility would be overloading Subscriber's constructor to also accept an instance of some simple delegate callback class, so the static usage would look something like:

```
class MyRosThing
{
  MyRosThing() :
    sub_cb_(&MyRosThing::twistCallback, this),
    sub_("cmd_drive", &sub_cb_) {}

  void twistCallback(const geometry_msgs::Twist&) {}

  ros::Callback<MyRosThing> sub_cb_;
  ros::Subscriber<geometry_msgs::Twist> sub_;
}
```

And the dynamic usage:

```
class MyRosThing
{
  MyRosThing()
  {
    sub_cb_ = new ros::Callback<MyRosThing>(&MyRosThing::twistCallback, this),
    sub_ = new ros::Subscriber<geometry_msgs::Twist>("cmd_drive", sub_cb_) {}
  }

  void twistCallback(const geometry_msgs::Twist&) {}

  ros::Callback* sub_cb_;
  ros::Subscriber* sub_;
}
```

Rosserial does no multiplexing (nor should it); realistically `MyRosThing` is going to be a singleton, so maybe it's not worth this complexity, and users with this type of architecture should simply implement a static wrapper which calls the singleton instance's method.

Thoughts on this, @mikeferguson, @PaulBouchier ?

**Comment by mikepurvis** [comment]
An alternative to rolling our own messy callback wrapper could be bundling or somehow otherwise supporting FastDelegate: http://www.codeproject.com/Articles/7150/Member-Function-Pointers-and-the-Fastest-Possible

Bundled, it could be fairly invisible to the user— the internal representation is a FastDelegate function pointer, and overloaded constructors allow either a FastDelegate or plain old function pointer to be passed.

**Comment by mikepurvis** [comment]
A simple step could be templating the CallbackT, eg changing this:

```
  template<typename MsgT>
  class Subscriber: public Subscriber_{
    public:
      typedef void(*CallbackT)(const MsgT&);
      MsgT msg;
```

to this:

```
  template<typename MsgT, typename CallbackT = void(*)(const MsgT&)>
  class Subscriber: public Subscriber_{
    public:
      MsgT msg;
```

Because of the default being supplied, I don't believe this affects the present behaviour at all, but does provide an option for the user to supply a delegating functor type.

If I have a chance to experiment further and it appears to bear fruit, I'll send a PR for review.

**Comment by kylerlaird** [comment]
I have been working around this with horrible macros.  I would love to have a cleaner way to do it.

**Comment by mikepurvis** [comment]
This is actually fixed as of #241 and #249.
